### ISSUE TITLE
про стирающие типы

### ISSUE THREAD

**Issue by DimaTrushin** [pulumboom/nn_from_scratch#11]
https://github.com/pulumboom/nn_from_scratch/blob/3983ee4744dec684d35c5fc929097373d4c4fd67/src/activation_functions/Sigmoid.h#L5

Идея стирающих типов заключается  в том, что тебе не надо будет ничего вот тут наследовать.

**Comment by pulumboom** [comment]
Вот здесь в вашей имплементации
![image](https://user-images.githubusercontent.com/61249560/232312338-a5d52723-40b6-4e3c-8fc7-4944d38c78a0.png)
все равно есть наследование, не понимаю как тогда можно избавиться от наследования. В докладе это избегается тем, что вместо методов используются функции, мне стоит сделать также или все-таки наследование оставить?

**Comment by DimaTrushin** [comment]
Нет, ты не понял в чем проблема. Наследование будет в любом случае, ты без виртуальных функций не сделаешь. Проблема не в этом. Наследование должно быть деталью имплементации, а у тебя наследование предоставлено пользователю. Вот два примера. Ты сделал следующее:
```cpp
struct I {
  void f() = 0;
};

struct A : I {
  void f() {}
};

struct Any {
  template<class T>
  Any(T&& data) : data_(std::make_unique<T>(std::forward<T>(data))) {}
  std::unique_ptr<I> data_;
};

int main() {
  Any x = A();
}
```
Проблема в том, что автор класса A должен унаследовать его от интерфейса I, и если этого не сделать, то ты этот объект не сможешь положить в Any. А вот другой пример
```cpp
struct I {
  void f() = 0;
};

template<class T>
struct Model : I {
  Model(T&& data) : data_(std::forward<T>())
  T data_;
};
struct Any {
  template<class T>
  Any(T&& data) : model_(std::make_unique<Model<T>>(std::forward<T>(data))) {}
  std::unique_ptr<I> model_;
};

struct A {
  void f() {}
};

int main() {
  Any x = A();
}
```
В этом примере, A не унаследовано от чего либо. Здесь в Any можно положить любой объект, у которого в интерфейсе есть метод void f().

Первое решение -- это просто заворачивание классического решения в unique_ptr. Но это не избавляет тебя от необходимости наследоваться от интерфейса. Второе решение -- позволяет хранить в классе Any что угодно, любой класс, даже не унаследованный от интерфейса. Ты продолбался с контруктором класса Any, ты не то строишь.
